File: backend/app/schemas/record.py
from pydantic import BaseModel
from datetime import datetime
from app.models.base import RecordType, PeriodType, Direction


class SimpleRecordCreate(BaseModel):
    name: str
    time: datetime
    period: PeriodType
    description: str | None = None


class PaymentRecordCreate(BaseModel):
    name: str
    description: str | None = None
    direction: Direction
    category: list[str] | None = None
    amount: float
    payment_method: list[str] | None = None
    period: PeriodType
    start_time: datetime
    end_time: datetime | None = None
    notes: str | None = None
    currency: str = "CNY"


class SimpleRecordUpdate(BaseModel):
    name: str
    time: datetime
    period: PeriodType
    description: str | None = None


class PaymentRecordUpdate(BaseModel):
    name: str
    description: str | None = None
    direction: Direction
    category: list[str] | None = None
    amount: float
    payment_method: list[str] | None = None
    period: PeriodType
    start_time: datetime
    end_time: datetime | None = None
    notes: str | None = None
    currency: str = "CNY"


class CustomRecordCreate(BaseModel):
    template_id: str
    custom_fields: dict


class RecordResponse(BaseModel):
    id: str
    type: RecordType
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


class SimpleRecordResponse(RecordResponse):
    type: RecordType = RecordType.SIMPLE
    name: str
    time: datetime
    period: PeriodType
    description: str | None
    next_occurrence: datetime | None


class PaymentRecordResponse(RecordResponse):
    type: RecordType = RecordType.PAYMENT
    name: str
    description: str | None
    direction: Direction
    category: list[str] | None
    amount: float
    payment_method: list[str] | None
    period: PeriodType
    start_time: datetime
    end_time: datetime | None
    notes: str | None
    currency: str
    next_occurrence: datetime | None

    class Config:
        orm_mode = True

    @classmethod
    def from_orm(cls, obj):
        import json

        data = {
            "id": obj.id,
            "type": obj.type,
            "created_at": obj.created_at,
            "updated_at": obj.updated_at,
            "name": obj.name,
            "description": obj.description,
            "direction": obj.direction,
            "amount": obj.amount,
            "period": obj.period,
            "start_time": obj.start_time,
            "end_time": obj.end_time,
            "notes": obj.notes,
            "currency": obj.currency,
            "next_occurrence": obj.next_occurrence,
        }
        if obj.category:
            try:
                data["category"] = (
                    json.loads(obj.category)
                    if isinstance(obj.category, str)
                    else obj.category
                )
            except:
                data["category"] = [obj.category]
        else:
            data["category"] = None

        if obj.payment_method:
            try:
                data["payment_method"] = (
                    json.loads(obj.payment_method)
                    if isinstance(obj.payment_method, str)
                    else obj.payment_method
                )
            except:
                data["payment_method"] = [obj.payment_method]
        else:
            data["payment_method"] = None

        return cls(**data)

File: frontend/src/types/record.ts
// Record types enum
export type RecordType = 'simple' | 'payment' | 'custom'

// Period types
export type PeriodType = 'week' | 'month' | 'quarter' | 'half-year' | 'year'

// Income/Expense direction
export type Direction = 'income' | 'expense'

// Currency
export type Currency = 'CNY' | 'USD'

// Currency display helpers
export const CURRENCY_SYMBOLS: Record<Currency, string> = {
  CNY: '¥',
  USD: '$',
}

export const CURRENCY_OPTIONS: {
  value: Currency
  label: string
  symbol: string
}[] = [
  { value: 'CNY', label: '人民币', symbol: '¥' },
  { value: 'USD', label: '美元', symbol: '$' },
]

// Period label mapping
export const PERIOD_LABELS: Record<PeriodType, string> = {
  week: '每周',
  month: '每月',
  quarter: '每季度',
  'half-year': '每半年',
  year: '每年',
}

export const PERIOD_OPTIONS: { value: PeriodType; label: string }[] = [
  { value: 'week', label: '周' },
  { value: 'month', label: '月' },
  { value: 'quarter', label: '季度' },
  { value: 'half-year', label: '半年' },
  { value: 'year', label: '年' },
]

export const DIRECTION_OPTIONS: { value: Direction; label: string }[] = [
  { value: 'income', label: '收入' },
  { value: 'expense', label: '支出' },
]

// Base record interface
export interface BaseRecord {
  id: string
  type: RecordType
  created_at: string
  updated_at: string
}

// Simple reminder record
export interface SimpleRecord extends BaseRecord {
  type: 'simple'
  name: string
  time: string
  period: PeriodType
  description?: string
  next_occurrence?: string
}

// Payment record
export interface PaymentRecord extends BaseRecord {
  type: 'payment'
  name: string
  description?: string
  direction: Direction
  category: string
  amount: number
  payment_method: string
  period: PeriodType
  start_time: string
  end_time?: string
  notes?: string
  currency: string
  next_occurrence?: string
}

// Union type for any record
export type CalendarRecord = SimpleRecord | PaymentRecord

// Support types
export interface Category {
  id: string
  name: string
  type: Direction
  color?: string
}

export interface PaymentMethod {
  id: string
  name: string
}

// Create DTOs
export interface SimpleRecordCreate {
  name: string
  time: string
  period: PeriodType
  description?: string
}

export interface PaymentRecordCreate {
  name: string
  description?: string
  direction: Direction
  category: string
  amount: number
  payment_method: string
  period: PeriodType
  start_time: string
  end_time?: string
  notes?: string
  currency?: string
}

// Update DTOs
export type SimpleRecordUpdate = SimpleRecordCreate
export type PaymentRecordUpdate = PaymentRecordCreate
